Testdata handling

cjauvin opened the pull request:
Je vais écrire en français ça va sortir plus fluidement. J'ai réfléchi aujourd'hui au mécanisme de "testdata" et je me suis demandé si on prévoyait vraiment un use case où un usager de RavenPy voudrait rouler seulement un sous-ensemble des tests. S'il veut rouler TOUS les tests, alors il me semble que c'est raisonnable (et plus simple) de lui demander de télécharger (ou git cloner) le repo `raven-testdata`, et c'est le point de départ de cette mécanique que je propose dans cette PR (incomplète, elle ne couvre pour le moment que `test_base.py`). Si l'usager ne veut pas rouler les tests, tout est ok et le repo reste petit. Il me semble qu'il n'y a que dans le cas où l'usager voudrait rouler juste quelques tests que la mécanique proposée dans `tutorial.py` est utile. Sinon il me semble que c'est complexe pour peu de gain.

Notez que pour la question du versioning et de la "fraicheur" du testdata, on peut laisser l'usager gérer ça lui-même avec git, qui est fait pour ça.

Finalement avec le modèle que je propose il reste le problème qu'il faut paramétrer le path du repo de testdata (hardcodé dans cette PR pour le moment).. ça pourrait être fait avec un git submodule peut-être, ou autre chose.

Ceci est le résultat de mes réflexions de la journée mais je considère qu'il est très possible que des détails du contexte plus large m'échappent.. si c'est le cas évidemment merci de me rappeler à l'ordre !

cjauvin commented:
J'ai oublié de dire : la raison à la base de cette idée est qu'en faisant des expériences, j'ai fini par dépasser la "API rate limit" de GH. Je me suis alors dit : pourquoi je voudrais accéder à GH à tous les coups? Si j'ai tout mon data localement et que j'en suis satisfait, je devrais pouvoir le faire. Mais évidemment on veut que le data reste dans un repo séparé.

cjauvin commented:
Aussi : la raison pourquoi j'accédais à GH à tous les coups est que j'utilisais le pattern `get_file(query_folder(..))` qui me semble être la seule possibilité d'utiliser un glob pattern pour le moment.

huard reviewed (approved):
One of the reasons for the remote test data repo was to facilitate writing tutorial notebooks. At the moment, running notebooks requires a copy of the TESTDATA folder, which implies raven is installed. WPS processes allows users to connect to remote services, without having to install raven locally. But to run those tutorial notebooks, we still need the local test data... 

Also, we want the same testdata for RavenPy and ravenWPS. 

So if we can have a solution that facilitates life for developers and allows user with no local installation to run notebooks, I'm happy.

For tutorial, I think the `get_file` access method will work. 
For tests, we could maybe use a pytest fixture to checkout the testdata repo.

cjauvin commented:
Hi all, I made a lot of changes here, but here's a summary: I have introduced a `--with-testdata` option that can be passed to `pip install` in order to download the testdata from the repo. I have modified all the tests accordingly. I plan to also create a `ravenpy download-testdata` command, that will do the same in the case where the testdata wasn't downloaded at install time, for convenience. Note that I also introduced a `--with-raven` option that works with the same logic, for the binaries.

I will merge this PR now, as I want to rapidly experiment with the behavior of all those mechanisms on PyPI.
